shared/annotation.py

```diff
@@ -642,12 +643,30 @@ def detect_and_remove_collisions(self, instance_list):
                 # Collision detected, we keep the newest instance by created time (which was order sorted).
                 # So this one is just to be deleted and not added to results.
                 logger.warning('Collision detected on {} instance id: {}'.format(inst.hash, inst.id))
+                logger.warning('hashes_dict: {}'.format(hashes_dict))
                 inst.soft_delete = True
                 inst.action_type = "from_collision"
                 self.session.add(inst)
 
         return result
 
+    def rehash_existing_instances(self, instance_list):
+        result = []
+        for instance in instance_list:
+            prev_hash = instance.hash
+            instance.hash_instance()
+            new_hash = instance.hash
+            if prev_hash != new_hash:
+                logger.warning('Warning: Hashing algorithm upgrade Instance ID: {} has changed \n from: {} \n to: {}'.format(
+                    instance.id,
+                    prev_hash,
+                    new_hash
+                ))
+                self.system_upgrade_hash_changes.append([prev_hash, new_hash])
+            result.append(instance)
+            self.session.add(instance)
```

Comment:
Should we only add to session if the change as occurred?  Context that we use that session add control in other contexts later in the saving process / if there's no changes there's no need to add it yet?

Reply:
Thanks Anthony! Added it here: https://github.com/diffgram/diffgram/pull/497/commits/ceccfd0d9250587c4e9e6854cc003dfb80594453